ENH add the possibility to pass array-like for max and min in IterativeImputer

narendramukherjee opened the pull request:
#### Reference Issues/PRs
Resolves #16349 

#### What does this implement/fix? 
The current implementation of `impute.IterativeImputer` only allows the specification of a single pair of (min, max) values to which all imputed features will be clipped. However, in many practical applications, different features might vary over different ranges. This PR adds functionality to the `IterativeImputer` to allow the specification of a different pair of min and max values for each feature to be imputed.

#### Any other comments?
These changes allow the following:
1. Leave the default `min_value` and `max_value` as `None` (which is the current default). In this case min and max values for all features will be set to `-np.inf` and `np.inf` respectively.
2. Allow the user to specify a single pair of numbers as min and max values for all features. In this case, the scalar `min_value` and `max_value` will be broadcast to float arrays of shape = number of features.
3. Allow the user to specify a list/array of min values and another list of max values, one value for each feature being imputed. The user can leave the min and max values of some features as `None` in these arrays in which case those values will be changed to `-np.inf` and `np.inf` while being cast as float arrays.

Since these changes do not modify the default behavior of `IterativeImputer`, I do not expect them to break any tests. However, I have no experience with testing while contributing to sklearn, so let me know if there is some work that I need to do wrt testing. 

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

narendramukherjee committed:
Use list of max and min values for the features

narendramukherjee committed:
Allow users to specify either: a) scalar, b) array/list, c) None, as min and max values. Changed documentation to reflect the same

narendramukherjee committed:
Broadcast scalar int min and max values to float array of shape num_features

narendramukherjee committed:
Move the broadcasting of min and max values to after the initial imputation so that the input X has already become an ndarray if it was a list (in response to failing test)

narendramukherjee committed:
Fixed PEP8 E127 in response to linting tests failure

narendramukherjee committed:
More fixes in response to linting errors

jnothman commented on sklearn/impute/_iterative.py:
```diff
@@ -581,6 +582,23 @@ def fit_transform(self, X, y=None):
+        # Convert min and max values to float arrays if array-like
+        # Else (i.e, if scalar) broadcast it to an array of shape=n_features
+        self._min_value = \
+            as_float_array(self.min_value, force_all_finite=False) \
```
Use parenthesised expression rather than `\` for line continuation, please.

narendramukherjee replied:
Done!

jnothman reviewed (commented):
Please add a test. Thanks!

narendramukherjee committed:
1. Added checks that raise ValueError if min_value >= max_value OR if min and max values do not have the same shape as n_features
2. Corrected formatting in response to review

narendramukherjee committed:
Add tests that:
1. Check if the min and max values as scalars, np.inf or None are cast to the right numpy arrays
3. Check if the right ValueError(s) are raised in fit_transform

narendramukherjee committed:
Minor linting error

narendramukherjee committed:
More linting fixes

narendramukherjee commented:
@jnothman 
I have made a number of changes in response to your comments:

1. Removed the use of "\" for line continuation

2. Added checks in `IterativeImputer.fit_transform` that see if a) `_min_value` < `_max_value` for each feature, and b) the `shape` of `_min_value` and `_max_value` is the same as the number of features. These checks raise `ValueError` if they fail - I have additionally put in a test case in (`test_impute.py/test_iterative_imputer_catch_min_max_error`) that checks if these `ValueError`(s) are raised properly.

3. Added a test in `test_impute.py` called `test_iterative_imputer_min_max_completion` which checks if all the variants of min and max value (scalar, list/tuple/array, None, -/+ inf) are being cast to numpy arrays properly.

Thanks again for your help and time, and let me know if you have additional comments.

jnothman commented on sklearn/impute/_iterative.py:
```diff
@@ -581,6 +582,36 @@ def fit_transform(self, X, y=None):
+        # Convert min and max values to float arrays if array-like
+        # Else (i.e, if scalar) broadcast it to an array of shape=n_features
+        self._min_value = (
+            as_float_array(self.min_value, force_all_finite=False)
```
Should we just use numpy.broadcast_to?

narendramukherjee replied:
`numpy.broadcast_to` is nifty, but it misses one critical condition - when `min_value` or `max_value` is `None`. `broadcast_to` converts `None` to an object type, and the usual `nan` checking (a couple of lines later) with `np.isnan` doesn't work. `as_float_array` from `sklearn` is appropriate for this situation in my opinion as `force_all_finite=False` takes care of these critical edge cases, and `None` is converted to `np.nan` and `inf` remains as `np.inf`. Since the default behavior of `IterativeImputer` is to specify `max_value` and `min_value` as `None`, this is an important case to take care of.

glemaitre replied:
```python
if np.isscalar(self._min_value):
    self._min_value = np.full(X.shape[1], self._min_value)
self._min_value = check_array(self._min_value, force_all_finite=False)
```

glemaitre replied:
IMO, we should not support `None` in arrays. Just numbers.

jnothman replied:
I'd handle the None case first and then use broadcast_to. I think that would be most readable. I am not sure why we support None to be honest! I'd be happy to change the default to +/-inf in a different pr

glemaitre replied:
I agree to support None as a scalar. I would not introduce the support for array and the next PR could make a deprecation (actually IterativeImputer is still experimental) for None toward +/-inf which will make thing consistent between array and scalar.

jnothman commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,59 @@ def test_iterative_imputer_catch_warning():
 
+@pytest.mark.parametrize("min_value, max_value",
+                         [(0, 100), (None, None), (-np.inf, np.inf),
+                          ((-5, 5, 10), (100, 200, 300)),
```
```suggestion
                          (([-5, 5, 10], [100, 200, 300]),
```

narendramukherjee replied:
Committed changes to the test resolving this!

jnothman commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,59 @@ def test_iterative_imputer_catch_warning():
+@pytest.mark.parametrize("min_value, max_value",
+                         [(0, 100), (None, None), (-np.inf, np.inf),
+                          ((-5, 5, 10), (100, 200, 300)),
+                          ((-5, None, 10), (100, 200, None))])
```
```suggestion
                          ([-5, None, 10], [100, 200, None])])
```

narendramukherjee replied:
Committed changes to the test resolving this

jnothman commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,59 @@ def test_iterative_imputer_catch_warning():
+    # Make a dict with the correct values that each pair
+    # of min and max values should be cast to
+    correct_vals_dict = (
+        {(0, 100): np.array([[0] * n_features,
```
Usually these values would be parameters to the test. Any reason not to?

narendramukherjee replied:
Committed changes to the test resolving this

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -111,13 +111,17 @@ class IterativeImputer(_BaseImputer):
-    min_value : float, default=None
-        Minimum possible imputed value. Default of ``None`` will set minimum
-        to negative infinity.
+    min_value : float or array-like, default=None
```
```suggestion
    min_value : float or array-like of shape (n_features,), default=None
```

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -111,13 +111,17 @@ class IterativeImputer(_BaseImputer):
-    max_value : float, default=None
-        Maximum possible imputed value. Default of ``None`` will set maximum
-        to positive infinity.
+    max_value : float or array-like, default=None
```
```suggestion
    max_value : float or array-like of shape (n_features,), default=None
```

narendramukherjee replied:
Where do we need to put this entry about what's new?

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -565,9 +569,6 @@ def fit_transform(self, X, y=None):
 
         self.imputation_sequence_ = []
 
-        self._min_value = -np.inf if self.min_value is None else self.min_value
```
Keep these lines.

narendramukherjee replied:
Done - but those lines are actually not needed as the use of `as_float_array` will convert `None` to `np.nan` anyways, and we later convert `np.nan` to -/+ `np.inf`. Your suggestion of using `check_array` doesn't work with integer arrays when we later try to convert `np.nan` in those arrays to `np.inf` - and we need to do this last step because the user can give an array-like input with `np.nan` or `None` in it which the condition above will miss. Actually, the condition above fails even for scalar `np.nan`.

Anyways, I have written up a detailed explanation for all that I tried below, have a look at that. The code could be more parsimonious, but I think we need a separate pull request to do that.

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -581,6 +582,36 @@ def fit_transform(self, X, y=None):
+        )
+        # Fill in any np.nan in min and max value arrays
+        # with -np.inf and np.inf respectively
+        self._min_value[np.isnan(self._min_value)] = -np.inf
```
I would remove this case

narendramukherjee replied:
That case is removed, but keeping `self._min_value[np.isnan(self._min_value)] = -np.inf` (and same for max value) as that handles the case of `np.nan` and `None` in array inputs (not scalars) correctly. Look at the details of all the options I tried (and which didn't work) in my explanation below.

glemaitre replied:
As mentioned below we should not support None inside an array so they will be no issue

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -581,6 +582,36 @@ def fit_transform(self, X, y=None):
+        self._min_value[np.isnan(self._min_value)] = -np.inf
+        self._max_value[np.isnan(self._max_value)] = np.inf
+
+        # Raise a ValueError if min value for any feature is >= its max value
```
No need for a comment, the code is self-explanatory.

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -581,6 +582,36 @@ def fit_transform(self, X, y=None):
+        # Raise a ValueError if min value for any feature is >= its max value
+        if not np.all(np.greater(self._max_value, self._min_value)):
+            raise ValueError(
+                "One (or more) features have min_value >= max_value")
```
```suggestion
                "One (or more) features have min_value >= max_value."
            )
```

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -581,6 +582,36 @@ def fit_transform(self, X, y=None):
+                "One (or more) features have min_value >= max_value")
+        # Raise a ValueError if _min_value and _max_value
+        # are not the same shape as n_features
+        if not (self._min_value.shape[0] == X.shape[1] and
```
This check should be done before the previous one.

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -581,6 +582,36 @@ def fit_transform(self, X, y=None):
+        if not np.all(np.greater(self._max_value, self._min_value)):
+            raise ValueError(
+                "One (or more) features have min_value >= max_value")
+        # Raise a ValueError if _min_value and _max_value
```
remove the comments

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -581,6 +582,36 @@ def fit_transform(self, X, y=None):
+        if not (self._min_value.shape[0] == X.shape[1] and
+                self._max_value.shape[0] == X.shape[1]):
+            raise ValueError(
+                "Number of min/max values not equal to number of features")
```
```suggestion
                f"'min_value' and 'max_values' should be of shape ({X.shape[1]},) "
                f"when an array-like is provided. Got {self._min_value.shape}, instead."
            )
```

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,59 @@ def test_iterative_imputer_catch_warning():
 
 
+@pytest.mark.parametrize("min_value, max_value",
+                         [(0, 100), (None, None), (-np.inf, np.inf),
```
I would not support None. Basically one have to build the array by hand and put `None` inside. Putting `np.inf` or `-np.inf` will be the same assle. It would simplify greatly the code on our side.

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,59 @@ def test_iterative_imputer_catch_warning():
+                         [(0, 100), (None, None), (-np.inf, np.inf),
+                          ((-5, 5, 10), (100, 200, 300)),
+                          ((-5, None, 10), (100, 200, None))])
+def test_iterative_imputer_min_max_completion(min_value, max_value):
```
```suggestion
def test_iterative_imputer_min_max_array_like(min_value, max_value):
```

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,59 @@ def test_iterative_imputer_catch_warning():
+    imputer.fit(X)
+    # Check if min_value and max_value have been cast to np arrays of
+    # shape = n_features
+    assert (isinstance(imputer._min_value, np.ndarray) and
```
make 2 different assert statement

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,59 @@ def test_iterative_imputer_catch_warning():
+    n_features = X.shape[1]
+    imputer = IterativeImputer(min_value=min_value, max_value=max_value)
+    imputer.fit(X)
+    # Check if min_value and max_value have been cast to np arrays of
```
no need for comment

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,59 @@ def test_iterative_imputer_catch_warning():
+                          ((-5, 5, 10), (100, 200, 300)),
+                          ((-5, None, 10), (100, 200, None))])
+def test_iterative_imputer_min_max_completion(min_value, max_value):
+    # Test for issue #16349 which adds the capability
```
Since this is not a regression, I think this is better to have 

```
# check that passing scalar or array-like works
```

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,59 @@ def test_iterative_imputer_catch_warning():
+                         [(0, 100), (None, None), (-np.inf, np.inf),
+                          ((-5, 5, 10), (100, 200, 300)),
+                          ((-5, None, 10), (100, 200, None))])
+def test_iterative_imputer_min_max_completion(min_value, max_value):
```
an additional test is to check that passing an array is equivalent to scalar.
additionally, passing array of `np.inf` and `-np.inf` is equivalent to `None`

narendramukherjee replied:
Those two tests are already part of the cases I have put in there. I have added comments to explain each test case, which should make it clearer.

glemaitre reviewed (changes requested):
We will an entry in what's new as well.

narendramukherjee committed:
Corrected the formatting of tests in response to reviews

narendramukherjee committed:
Update sklearn/impute/_iterative.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

narendramukherjee committed:
Update sklearn/impute/_iterative.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jnothman reviewed (commented):
Please address @glemaitre's remaining comments. This is looking pretty good.

Please add an entry to the change log at `doc/whats_new/v0.23.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`

narendramukherjee commented:
@glemaitre @jnothman 
I spent quite a lot of time trying to do the following (but failed):
1. Remove the usage on `None` and put `-np.inf` and `np.inf` as defaults. **That breaks a large number of tests**, so I think that change affects all the imputer modules and will need a separate PR
2. Use `np.broadcast_to` and/or `check_array`. Neither covers all the cases we are looking at: `np.broadcast_to` gives a read-only view as an array, and it is not possible to replace any remaining `np.nan` in that array. `check_array` fails with integer arrays and is not able to replace `np.nan` in integer arrays with -/+`np.inf`. Overall, however clean the code might look with their usage, these options do not cover all edge cases correctly.

So eventually, I am going ahead with the following strategy, which incorporates some (but not all) of your input:
1. Check if `min_value` or `max_value` are `None`. Replace with -/+ `np.inf` if yes. This check will miss `np.nan` which will be dealt with in the next steps.
2. Check if `min_value` or `max_value` are scalar. Convert them to arrays (filled with that scalar value) of shape=n_features.
3. Convert these arrays to float arrays using `as_float_array`. If the user put in `None` for a subset of the features, that wouldn't be caught by step 1. In this step, those values will be converted to `np.nan`
4. Replace all `np.nan` in the arrays with -/+ `np.inf`. This will catch any `np.nan` that escapes step 1 or `np.nan` which appear in step 3.

I really couldn't figure out a more parsimonious way of doing this, and would prefer going ahead with this strategy and not spending even more time on this.

narendramukherjee committed:
Changes in response to review comments

narendramukherjee committed:
Linting fix

narendramukherjee committed:
Added description to what's new

narendramukherjee commented:
@glemaitre @jnothman 
I have committed my latest changes and responded to all the review comments. I have added a description of what's new to `doc/whats_new/v0.23.rst`. Let me know what you guys think.

jnothman commented on doc/whats_new/v0.23.rst:
```diff
@@ -116,7 +116,13 @@ Changelog
+:mod:`sklearn.impute`
+.....................
+
+- |Enhancement| :class:`impute.IterativeImputer` accepts both scalar and array-like inputs for
```
Tell it to the user's perspective: say that we now allow max and min to be specified per feature

narendramukherjee replied:
Done!

jnothman reviewed (commented):
I'd feel more comfortable if we also had a test that these per-feature settings had an effect on the overall imputation. Otherwise lgtm

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,42 +971,39 @@ def test_iterative_imputer_catch_warning():
-                          ((-5, 5, 10), (100, 200, 300)),
-                          ((-5, None, 10), (100, 200, None))])
-def test_iterative_imputer_min_max_completion(min_value, max_value):
+@pytest.mark.parametrize("min_value, max_value, correct_output",
```
Here you can add a new line, it will be more readable for the rest of the code:

```python
@pytest.mark.parametrize(
    "min_value, max_value, correct_output",
    [(0, 100, np.array([[0] * 3, [100] * 3])),
     (...),
     ...]
)
```

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,42 +971,39 @@ def test_iterative_imputer_catch_warning():
-                       imputer._max_value)
+            (imputer._max_value.shape[0] == X.shape[1]))
+    # Check if the imputer has correctly processed the min and max values
+    assert_array_equal(correct_output[0, :], imputer._min_value)
```
if we are going to have `inf`, the dtype of the array for comparison will be float and therefore we should use `assert_allclose` instead of `assert_array_equal`.

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -111,13 +111,15 @@ class IterativeImputer(_BaseImputer):
-        to negative infinity.
+    min_value : float or array-like of shape (n_features,), default=None.
+        Minimum possible imputed value. Broadcast to shape (n_features,) if
+        scalar. Expects shape (n_features,) if array-like, one min value for
```
I would turn the sentence around `If array-like, the array should be of shape ...

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -111,13 +111,15 @@ class IterativeImputer(_BaseImputer):
+    min_value : float or array-like of shape (n_features,), default=None.
+        Minimum possible imputed value. Broadcast to shape (n_features,) if
+        scalar. Expects shape (n_features,) if array-like, one min value for
+        each feature. Default of None is converted to -np.inf.
```
```suggestion
        each feature. `None` (default) is converted to `-np.inf`.
```

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -111,13 +111,15 @@ class IterativeImputer(_BaseImputer):
-        Maximum possible imputed value. Default of ``None`` will set maximum
-        to positive infinity.
+    max_value : float or array-like of shape (n_features,), default=None.
+        Maximum possible imputed value. Broadcast to shape (n_features,) if
```
Same comment than above

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -581,6 +580,37 @@ def fit_transform(self, X, y=None):
             self.n_iter_ = 0
             return super()._concatenate_indicator(Xt, X_indicator)
 
+        # If min or max value are (scalar) None, convert them to - or + inf
```
I agree with Joel. Also remove all comments. The code is self-explanatory. We try to limit comment for aspect which is not trivial by just reading the code. If the operation is not obvious, it might mean that we need to change the code then :)

glemaitre replied:
To be more explicit, I think that `_validate_limit` should only validate a single limit. We have the exact same sequence of operation for min and max. The differences should be triggered by the parameter passed to this method.

glemaitre replied:
So the function should be:

```python
@staticmethod
    def _validate_limit(limit, limit_type, n_features):
        limit_bound = np.inf if limit_type == "max" else -np.inf
        limit = limit_bound if limit is None else limit
        if np.isscalar(limit):
            limit = np.full(n_features, limit)
        limit = check_array(
            limit, force_all_finite=False, copy=False, ensure_2d=False
        )
        if not limit.shape[0] == n_features:
            raise ValueError(
            f"'{limit_type}_value' should be of shape ({n_features},) when "
            f"an array-like is provided. Got {limit.shape}, instead."
        )
        return limit
```

narendramukherjee replied:
Done - thank you for fleshing out the method :)

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -581,6 +580,37 @@ def fit_transform(self, X, y=None):
+        if np.isscalar(self._max_value):
+            self._max_value = np.full(X.shape[1], self._max_value)
+        # If there were any None values in above arrays, convert them to np.nan
+        self._min_value = as_float_array(self._min_value,
```
do not use `as_float_array`. Use `check_array` instead.

narendramukherjee replied:
All these pulled into the `_validate_limit` method as you suggested :)

glemaitre reviewed (commented):
The failing test is only because you need to merge master into your branch.
I want to check a bit more the PR but this is a first round of comments.

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,58 @@ def test_iterative_imputer_catch_warning():
+                          ([-5, 5], [100, 200, 0])])
+def test_iterative_imputer_catch_min_max_error(min_value, max_value):
+    # check that passing scalar or array-like
+    # for min_value and max_value in IterativeImputer works
```
We don't need comments as well. For instance here we cannot know the the case that you are testing in practise because you are not matching the error message.

Thus if you are doing `pytest.raises(ValueError, match="err_msg")`, you can pass the expected (partial error messgae). `min_value should be of shape` or `One (or more) ...`. Then the test will be self explenatory.

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,58 @@ def test_iterative_imputer_catch_warning():
 
 
+@pytest.mark.parametrize("min_value, max_value, correct_output",
+                         [(0,  # scalars cast to the correct arrays?
```
instead of putting a comment here we can pass a parameter `ids` to name the different try. It will be easier when running the tests:

```python
@pytest.mark.parametrize(
    "min_value, max_value, correct_output",
    ...,
    ids=["scalars", "list", "None-default", "inf", ...]
)
```

narendramukherjee replied:
Done!

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,58 @@ def test_iterative_imputer_catch_warning():
+                          ([-5, 5, 10],  # vectors processed properly?
+                           [100, 200, 300],
+                           np.array([[-5, 5, 10], [100, 200, 300]])),
+                          ([-5, None, 10],  # vectors of numbers+None
```
We should not support the case that we have `None` when this is an array.

narendramukherjee replied:
Done!

narendramukherjee replied:
Done!

glemaitre reviewed (commented):
I would still add an additional test to check that `None` will be equivalent to pass array of `np.inf`.

glemaitre commented:
@narendramukherjee It starts to look good and we are converging toward the solution.

narendramukherjee committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

narendramukherjee committed:
Change what's new in response to review comments

narendramukherjee committed:
Changes in response to review comments

narendramukherjee commented:
Still working on this, will update when I have addressed all the reviews and made commits (in the next couple of days). Thanks for your time and help!

narendramukherjee committed:
Response to review comments

narendramukherjee committed:
Linting fix

narendramukherjee committed:
Added a test that compares imputation with min/max being None with inf and scalar with vector

narendramukherjee commented:
@glemaitre @jnothman 
I have addressed/responded to all your comments and incorporated all the changes you suggested. I have additionally included 1 more test in `test_iterative_imputer_min_max_array_like_imputation`: this checks if the imputation is the same a) whether you give `None` or `np.inf` as min/max values, and b) whether you give a scalar or its expanded vector as min/max values.

Let me know what you think, and thanks for your help :)

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,75 @@ def test_iterative_imputer_catch_warning():
+        [np.nan, 1, 10, 1]])
+    imputer1 = IterativeImputer(min_value=min_max_1[0],
+                                max_value=min_max_1[1],
+                                random_state=rng)
```
```suggestion
                                random_state=0)
```

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,75 @@ def test_iterative_imputer_catch_warning():
+                                random_state=rng)
+    imputer2 = IterativeImputer(min_value=min_max_2[0],
+                                max_value=min_max_2[1],
+                                random_state=rng)
```
```suggestion
                                random_state=0)
```

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,75 @@ def test_iterative_imputer_catch_warning():
+    ids=["None-vs-inf", "Scalar-vs-vector"])
+def test_iterative_imputer_min_max_array_like_imputation(min_max_1, min_max_2):
+    # Test that None/inf and scalar/vector give the same imputation
+    rng = np.random.RandomState(0)
```
```suggestion
```

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -971,6 +971,75 @@ def test_iterative_imputer_catch_warning():
+                                              correct_output):
+    # check that passing scalar or array-like
+    # for min_value and max_value in IterativeImputer works
+    X = np.random.random((10, 3))
```
```suggestion
    X = np.random.RandomState(0).random((10, 3))
```

glemaitre reviewed (approved):
Apart of the last changes, it looks good.

narendramukherjee committed:
Update sklearn/impute/tests/test_impute.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

narendramukherjee committed:
Update sklearn/impute/tests/test_impute.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

narendramukherjee committed:
Update sklearn/impute/tests/test_impute.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

narendramukherjee committed:
Update sklearn/impute/tests/test_impute.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

narendramukherjee committed:
Merge branch 'master' into master

narendramukherjee commented:
Thanks for all your help @glemaitre ! I have incorporated your suggested changes, let me know when the changes can be merged into master of scikit-learn.

glemaitre committed:
Update test_impute.py

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -970,6 +970,74 @@ def test_iterative_imputer_catch_warning():
+    ids=["None-vs-inf", "Scalar-vs-vector"])
+def test_iterative_imputer_min_max_array_like_imputation(min_max_1, min_max_2):
+    # Test that None/inf and scalar/vector give the same imputation
+    X_train = np.array([
```
Uhm this is a weird case here. Basically, we want to have some missing value in the training set (you can put some in the different columns and in the test the same thing (but not a full NaN column)

glemaitre replied:
I quickly change the values to check that the test pass. If this is fine, we will need a second review for a second check.

glemaitre committed:
Update test_impute.py

glemaitre commented:
@jnothman do you want to have a last look at it?

glemaitre committed:
Update test_impute.py

rth reviewed (approved):
LGTM as well, thanks @narendramukherjee !
